subsys/bluetooth/mesh/provisioner.c

```diff
@@ -0,0 +1,767 @@
+}
+
+static void send_invite(void)
+{
+	PROV_BUF(inv, 2);
+
+	BT_DBG("");
+
+	bt_mesh_prov_buf_init(&inv, PROV_INVITE);
+	net_buf_simple_add_u8(&inv, prov_device.attention_duration);
+
+	bt_mesh_prov_link.conf_inputs[0] = prov_device.attention_duration;
+
+	if (bt_mesh_prov_send(&inv, NULL)) {
+		BT_ERR("Failed to send invite");
+		return;
+	}
+
+	bt_mesh_prov_link.expect = PROV_CAPABILITIES;
+}
+
+static void start_sent(int err, void *cb_data)
+{
+	if (!bt_pub_key_get()) {
+		atomic_set_bit(bt_mesh_prov_link.flags, WAIT_PUB_KEY);
+		BT_WARN("Waiting for local public key");
+	} else {
+		send_pub_key();
+	}
+}
+
+static void send_start(uint8_t method, uint8_t action, uint8_t size)
+{
+	BT_DBG("");
+
+	PROV_BUF(start, 6);
+
+	const uint8_t *data = &bt_mesh_prov_link.conf_inputs[1 + 3];
+
+	bt_mesh_prov_buf_init(&start, PROV_START);
+	net_buf_simple_add_u8(&start, PROV_ALG_P256);
+
+	if (atomic_test_bit(bt_mesh_prov_link.flags, REMOTE_PUB_KEY) &&
+	    *data == PUB_KEY_OOB) {
+		net_buf_simple_add_u8(&start, PUB_KEY_OOB);
+		atomic_set_bit(bt_mesh_prov_link.flags, OOB_PUB_KEY);
+	} else {
+		net_buf_simple_add_u8(&start, PUB_KEY_NO_OOB);
+	}
+
+	net_buf_simple_add_u8(&start, bt_mesh_prov_link.oob_method);
+
+	net_buf_simple_add_u8(&start, bt_mesh_prov_link.oob_action);
+
+	net_buf_simple_add_u8(&start, bt_mesh_prov_link.oob_size);
+
+	memcpy(&bt_mesh_prov_link.conf_inputs[12], &start.data[1], 5);
+
+	if (bt_mesh_prov_auth(method, action, size) < 0) {
+		BT_ERR("Invalid authentication method: 0x%02x; "
+		       "action: 0x%02x; size: 0x%02x", method,
+		       action, size);
+		return;
+	}
+
+	if (bt_mesh_prov_send(&start, start_sent)) {
+		BT_ERR("Failed to send Provisioning Start");
+		return;
+	}
+}
+
+static void prov_check_method(struct bt_mesh_prov_capabilities *caps)
+{
+	uint8_t method = AUTH_METHOD_NO_OOB, action = 0x00, size = 0x00;
+
+	if (bt_mesh_prov_link.oob_method == AUTH_METHOD_STATIC) {
+		if (!caps->static_oob) {
+			BT_WAIN("Device not support OOB static authentication provisioning");
+			goto failed;
+		}
+
+		method = AUTH_METHOD_STATIC;
+	} else if (bt_mesh_prov_link.oob_method == AUTH_METHOD_INPUT) {
+		if (bt_mesh_prov_link.oob_size > caps->input_size) {
+			BT_WAIN("The required input length (0x%02x) "
+				"exceeds the device capacity (0x%02x)",
+				bt_mesh_prov_link.oob_size, caps->input_size);
+			goto failed;
+		}
+
+		if (!(BIT(bt_mesh_prov_link.oob_action) & caps->input_actions)) {
+			BT_WAIN("The required input action (0x%02x) "
+				"not supported by the device (0x%02x)",
+				bt_mesh_prov_link.oob_action, caps->input_actions);
+			goto failed;
+		}
+
+		if (bt_mesh_prov_link.oob_action == INPUT_OOB_STRING) {
+			if (!bt_mesh_prov->output_string) {
+				BT_WAIN("Not support output string");
+				goto failed;
+			}
+
+			action = OUTPUT_OOB_STRING;
+		} else {
+			action = find_msb_set(bt_mesh_prov->output_actions &
+					      (~BT_MESH_DISPLAY_STRING));
+			if (!action || !bt_mesh_prov->output_number) {
```

Comment:
Still unresolved: Only check whether the callback exists instead, since that's the only thing we actually need to execute this.

Here and below.

Reply:
Why not check the output action? If Device only support input number, and provisioner only support output string, this will be error.

Reply:
because the oob actions in the prov struct represent the authentication methods that are supported on the device in the provisionee role, just like the static oob value is the local static oob value. Whether this device supports blinking, pushing and so on isn't really relevant when acting in the provisioner role

Reply:
Okay,The specification indeed doesn't require how provisioner to input and output.